rts/c/reader-bin.h

```diff
@@ -0,0 +1,298 @@
+    }
+    return read_str_int64(dest);
+}
+
+static int read_float(void* dest) {
+    if (read_is_binary()) {
+        read_bin_ensure_scalar(FUTHARK_FLOAT32);
+        if (IS_BIG_ENDIAN) {
+            return read_be_4byte(dest);
+        } else {
+            return read_le_4byte(dest);
+        }
+    }
+    return read_str_float(dest);
+}
+
+static int read_double(void* dest) {
+    read_bin_ensure_scalar(FUTHARK_FLOAT64);
+    if (read_is_binary()) {
+        if (IS_BIG_ENDIAN) {
+            return read_be_8byte(dest);
+        } else {
+            return read_le_8byte(dest);
+        }
+    }
+    return read_str_double(dest);
+}
+
+static int read_bool(void* dest) {
+    if (read_is_binary()) {
+        read_bin_ensure_scalar(FUTHARK_BOOL);
+        return read_byte(dest);
+    }
+    return read_str_bool(dest);
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// General array interface
+////////////////////////////////////////////////////////////////////////////////
+
+static int read_array(primtype_t expected_type, int64_t elem_size, int (*elem_reader)(void*),
+                      const char *type_name, void **data, int64_t *shape, int64_t dims) {
+    if (!read_is_binary()) {
+        return read_str_array(elem_size, elem_reader, type_name, data, shape, dims);
+    }
+
+    // now we know it is binary :)
+    int ret;
+
+    int8_t bin_dims;
+    ret = read_byte(&bin_dims);
+    if (ret != 0) { panic(1, "binary-input: Couldn't get dims.\n"); }
+
+    if (bin_dims != dims) {
+        panic(1, "binary-input: Expected %i dimensions, but got array with %i dimensions.\n",
+              dims, bin_dims);
+    }
+
+    primtype_t bin_type_enum = read_bin_read_type_enum();
+    const primtype_info_t bin_primtype = FUTHARK_PRIMTYPES[bin_type_enum];
+    if (expected_type != bin_type_enum) {
+        panic(1, "binary-input: Expected %iD-array with element type '%s' but got %iD-array with element type '%s'.\n",
+              dims, FUTHARK_PRIMTYPES[expected_type].type_name, dims, bin_primtype.type_name);
+    }
+
+    if (elem_size != bin_primtype.size) {
+        panic(1, "binary-input: The RTS expected type %s to use %i bytes per element, but the call to `read_array` tells me to use %i bytes per element.\n",
+              bin_primtype.type_name, bin_primtype.size, elem_size);
+    }
+
+    int64_t elem_count = 1;
+    for (int i=0; i<dims; i++) {
+        int64_t bin_shape;
+        ret = IS_BIG_ENDIAN ? read_be_8byte(&bin_shape) : read_le_8byte(&bin_shape);
+        if (ret != 0) { panic(1, "binary-input: Couldn't read size for dimension %i of array.\n", i); }
+        elem_count *= bin_shape;
+        shape[i] = bin_shape;
+    }
+
+    void* tmp = realloc(*data, elem_count * elem_size);
+    if (tmp == NULL) {
+        panic(1, "binary-input: Failed to allocate array of size %i.\n",
+              elem_count * elem_size);
+    }
+    *data = tmp;
+
+    if (IS_BIG_ENDIAN && elem_size != 1) {
```

Comment:
Instead of this, why not just always fread() and then flip the bytes if on a big-endian platform?  I think your control flow is too complicated this way.